pytest_mpl/kernels.py

```diff
@@ -0,0 +1,223 @@
+"""
+This module contains the supported hashing kernel implementations.
+
+"""
+import hashlib
+from abc import ABC, abstractmethod
+
+import imagehash
+from PIL import Image
+
+#: The default hamming distance bit tolerance for "similar" imagehash hashes.
+DEFAULT_HAMMING_TOLERANCE = 2
+
+#: The default imagehash hash size (N), resulting in a hash of N**2 bits.
+DEFAULT_HASH_SIZE = 16
+
+#: Level of image detail (high) or structure (low) represented by phash .
+DEFAULT_HIGH_FREQUENCY_FACTOR = 4
+
+#: Registered kernel names.
+KERNEL_SHA256 = "sha256"
+KERNEL_PHASH = "phash"
+
+__all__ = [
+    "DEFAULT_HAMMING_TOLERANCE",
+    "DEFAULT_HASH_SIZE",
+    "DEFAULT_HIGH_FREQUENCY_FACTOR",
+    "KERNEL_PHASH",
+    "KERNEL_SHA256",
+    "KernelPHash",
+    "KernelSHA256",
+    "kernel_factory",
+]
+
+
+class Kernel(ABC):
+    """
+    Kernel abstract base class (ABC) which defines a simple common kernel API.
+
+    """
+
+    def __init__(self, plugin):
+        # Containment of the plugin allows the kernel to cherry-pick required state.
+        self._plugin = plugin
```

Comment:
It's a weak contract, but the deal here is that a `kernel` is forbidden to alter the contents of the `self._plugin` - that's just looking for trouble.

So this is a read-only private property. I didn't want to over-engineer that contract to protect ourselves from ourselves, as I didn't see that as worthwhile... but you may think otherwise.